File: blog/2025/10/extract_emmc_bin.py
import sys
import os

## VARS ################################

B  = 2**00
KB = 2**10
MB = 2**20
GB = 2**30

BLOCK_SIZE = 0x200
HEADER_SIZE = 0x200000 # Header size is always 2MB

TYPE_PARTITION = 22592

OFFSET_NAME = 0x10
OFFSET_BASE = 0x08
OFFSET_SIZE = 0X0C

## FUNCTIONS ##############################

def createDirectory(dir):
	if not os.path.exists(dir): # if the directory does not exist
		os.makedirs(dir) # make the directory
	else: # the directory exists
		#removes all files in a folder
		for the_file in os.listdir(dir):
			file_path = os.path.join(dir, the_file)
			try:
				if os.path.isfile(file_path):
					os.unlink(file_path) # unlink (delete) the file
			except e:
				print (e)

def loadPart(file, offset, size):
	with open(file, 'rb') as f:
		f.seek(offset)
		return f.read(size)

def copyPart(src, dest, offset, size, bufsize = 16 * MB, append = False):
	
	if not os.path.exists(dest):
		append = False
	
	with open(src, 'rb') as f1:
		f1.seek(offset)
		with open(dest, 'ab' if append else 'wb') as f2:
			while size:
				chunk = min(bufsize, size)
				data = f1.read(chunk)
				f2.write(data)
				size -= chunk

## PROGRAMM ##############################

disk = False

# Parse args
if len(sys.argv) == 1:
	print ("Usage: unpack.py <emmc.bin> <block size [default: 512]> <output folder [default: ./emmc/]>\n")
	disk = input("Do you want open disk? (choose letter): ")
	print()
	if len(disk) == 1:
		# on Linux it's like /dev/sda1
		inputFile = "\\\\.\\"+disk.upper()+":"
	else:
		print("Nothing to do - exiting")
		quit()

if not disk:
	inputFile = sys.argv[1]

if not disk and not os.path.exists(inputFile):
	print ("No such file: {}".format(inputFile))
	quit()

if len(sys.argv) >= 3:
	BLOCK_SIZE = int(sys.argv[2])

if len(sys.argv) == 4:
	outputDirectory = sys.argv[3]
else:
	outputDirectory = 'emmc'

# Create output directory
createDirectory(outputDirectory)

# Find header
# Header size is 2MB
if disk:
	print ("Direct disk read mode - "+inputFile)
else:
	print ("\nFile size:  " + format(os.path.getsize(inputFile),",")+" Bytes")
	
print ("Block size: " + str(BLOCK_SIZE)+" Bytes")
print ("\nPartitions:\n")

header = loadPart(inputFile, 0, HEADER_SIZE)
copyPart(inputFile, os.path.join(outputDirectory, "~header"), 0, HEADER_SIZE)

partitions = []
n = 0

for x in range(0, len(header), BLOCK_SIZE):
	type = int.from_bytes(header[x:x+2], byteorder ='little')
	if type == TYPE_PARTITION:
		n += 1
		name = header[x+OFFSET_NAME:x+OFFSET_NAME+20].decode('ascii').rstrip('\x00')
		base = int.from_bytes(header[x+OFFSET_BASE:x+OFFSET_BASE+4], byteorder ='little')
		size = int.from_bytes(header[x+OFFSET_SIZE:x+OFFSET_SIZE+4], byteorder ='little')
		full_size = size * BLOCK_SIZE // 1024
		
		if full_size > 1024:
			full_size = str(full_size // 1024) + " MB"
		else:
			full_size = str(full_size) + " KB"
			
		print(str(n).rjust(2,'0') + ": " + name.ljust(30) + " " + str(str(base)+":"+str(size)).ljust(25) + " " + full_size )
		partitions.append({'name':name, 'base':base, 'size':size});

if len(partitions) < 1:
	print ("\nPartions are not found\n")
	quit()
	
print ("\nExtracting...\n")

all = str(len(partitions))
n = 1

for p in partitions:
	outFile = os.path.join(outputDirectory, p['name']+".bin")
	print("Saving ("+str(n).rjust(2,'0')+"/"+all+") "+(p['name']+" ["+str(p['base']) + ":" + str(p['size']) + "]").ljust(40)+" => "+outFile)
	copyPart(inputFile, outFile, p['base']*BLOCK_SIZE, p['size']*BLOCK_SIZE)
	n += 1

File: internal/extract-emmc-bin/extract-emmc-bin.go
package main

import (
   "bufio"
   "encoding/binary"
   "fmt"
   "io"
   "log"
   "os"
   "path/filepath"
   "strconv"
   "strings"
)

// VARS
const (
   B             = 1 << 0
   KB            = 1 << 10
   MB            = 1 << 20
   GB            = 1 << 30
   HEADERSIZE    = 2 * MB // Header size is always 2MB
   TYPEPARTITION = 22592
   OFFSETNAME    = 0x10
   OFFSETBASE    = 0x08
   OFFSETSIZE    = 0x0C
)

var BLOCKSIZE int64 = 0x200

// Partition struct to hold partition information
type Partition struct {
   Name string
   Base int64
   Size int64
}

// createDirectory ensures a directory exists and is empty.
func createDirectory(dir string) {
   if _, err := os.Stat(dir); os.IsNotExist(err) {
      if err := os.MkdirAll(dir, 0755); err != nil {
         log.Fatalf("Failed to create directory: %v", err)
      }
   } else {
      files, err := os.ReadDir(dir)
      if err != nil {
         log.Fatalf("Failed to read directory: %v", err)
      }
      for _, file := range files {
         filePath := filepath.Join(dir, file.Name())
         if err := os.Remove(filePath); err != nil {
            log.Printf("Failed to remove file: %v", err)
         }
      }
   }
}

// =================================================================
// UPDATED copyPart FUNCTION
// It now accepts an already-open file handle (*os.File)
// =================================================================
func copyPart(srcFile *os.File, dest string, offset int64, size int64) {
   // Seek to the correct offset for the partition in the source file
   if _, err := srcFile.Seek(offset, io.SeekStart); err != nil {
      log.Fatalf("Failed to seek in source file: %v", err)
   }

   // Create the destination file for the partition
   destFile, err := os.Create(dest)
   if err != nil {
      log.Fatalf("Failed to create destination file: %v", err)
   }
   defer destFile.Close()

   // Use io.Copy with an io.LimitReader to efficiently copy the exact number of bytes
   if _, err := io.Copy(destFile, io.LimitReader(srcFile, size)); err != nil {
      log.Fatalf("Failed to copy data: %v", err)
   }
}

func main() {
   var inputFile string
   var outputDirectory = "emmc"

   if len(os.Args) < 2 {
      fmt.Println("Usage: unpack <emmc.bin> <block size [default: 512]> <output folder [default: ./emmc/]>")
      fmt.Print("Do you want open disk? (choose letter): ")
      reader := bufio.NewReader(os.Stdin)
      disk, _ := reader.ReadString('\n')
      disk = strings.TrimSpace(disk)
      fmt.Println()

      if len(disk) == 1 {
         inputFile = `\\.\` + strings.ToUpper(disk) + ":"
      } else {
         fmt.Println("Nothing to do - exiting")
         return
      }
   } else {
      inputFile = os.Args[1]
   }

   if len(os.Args) >= 3 {
      bs, err := strconv.ParseInt(os.Args[2], 10, 64)
      if err != nil {
         log.Fatalf("Invalid block size: %v", err)
      }
      BLOCKSIZE = bs
   }

   if len(os.Args) == 4 {
      outputDirectory = os.Args[3]
   }

   createDirectory(outputDirectory)

   // =================================================================
   // IMPROVEMENT: Open the source file ONCE here
   // =================================================================
   srcFile, err := os.Open(inputFile)
   if err != nil {
      log.Fatalf("Failed to open input file %s: %v", inputFile, err)
   }
   // Ensure the file is closed when the program exits
   defer srcFile.Close()

   if strings.HasPrefix(inputFile, `\\.\`) {
      fmt.Printf("Direct disk read mode - %s\n", inputFile)
   } else {
      fileInfo, err := srcFile.Stat()
      if err != nil {
         log.Fatalf("Failed to get file info: %v", err)
      }
      fmt.Printf("\nFile size:  %d Bytes\n", fileInfo.Size())
   }

   fmt.Printf("Block size: %d Bytes\n", BLOCKSIZE)
   fmt.Println("\nPartitions:\n")

   // Read the header from the already-open file
   header := make([]byte, HEADERSIZE)
   if _, err := srcFile.ReadAt(header, 0); err != nil {
      log.Fatalf("Failed to read header: %v", err)
   }
   
   // Save the header file
   copyPart(srcFile, filepath.Join(outputDirectory, "~header"), 0, HEADERSIZE)


   var partitions []Partition
   for x := 0; x < len(header); x += int(BLOCKSIZE) {
      partitionType := binary.LittleEndian.Uint16(header[x : x+2])
      if partitionType == TYPEPARTITION {
         nameBytes := header[x+OFFSETNAME : x+OFFSETNAME+20]
         name := strings.TrimRight(string(nameBytes), "\x00")
         base := binary.LittleEndian.Uint32(header[x+OFFSETBASE : x+OFFSETBASE+4])
         size := binary.LittleEndian.Uint32(header[x+OFFSETSIZE : x+OFFSETSIZE+4])
         fullSize := int64(size) * BLOCKSIZE / 1024
         var fullSizeStr string
         if fullSize > 1024 {
            fullSizeStr = fmt.Sprintf("%d MB", fullSize/1024)
         } else {
            fullSizeStr = fmt.Sprintf("%d KB", fullSize)
         }
         fmt.Printf("%02d: %-30s %-25s %s\n", len(partitions)+1, name, fmt.Sprintf("%d:%d", base, size), fullSizeStr)
         partitions = append(partitions, Partition{Name: name, Base: int64(base), Size: int64(size)})
      }
   }

   if len(partitions) < 1 {
      fmt.Println("\nPartitions are not found")
      return
   }

   fmt.Println("\nExtracting...\n")

   for i, p := range partitions {
      outFile := filepath.Join(outputDirectory, p.Name+".bin")
      fmt.Printf("Saving (%02d/%d) %-40s => %s\n", i+1, len(partitions), fmt.Sprintf("%s [%d:%d]", p.Name, p.Size, p.Base), outFile)
      // =================================================================
      // Pass the open file handle, not the filename string
      // =================================================================
      copyPart(srcFile, outFile, p.Base*BLOCKSIZE, p.Size*BLOCKSIZE)
   }
}
